File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKScene.java
package br.com.insanitech.spritekit;

import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnTouchListener;
import br.com.insanitech.spritekit.opengl.renderer.GLRenderer;

public class SKScene extends SKEffectNode implements OnTouchListener {
	public enum SKSceneScaleMode {
		SKSceneScaleModeFill, SKSceneScaleModeAspectFill, SKSceneScaleModeAspectFit, SKSceneScaleModeResizeFill
	}

	private SKSceneScaleMode scaleMode = SKSceneScaleMode.SKSceneScaleModeAspectFit;
	private SKColor backgroundColor = SKColor.whiteColor();
	private SKPoint anchorPoint = new SKPoint();
	private SKSize size = new SKSize();
	private SKView view;

	public static SKScene sceneWithSize(SKSize size) {
		return new SKScene(size);
	}

	public SKScene(SKSize size) {
		this.size.assignByValue(size);
	}

	public SKSceneScaleMode getScaleMode() {
		synchronized (this) {
			return scaleMode;
		}
	}

	public void setScaleMode(SKSceneScaleMode mode) {
		synchronized (this) {
			scaleMode = mode;
		}
	}

	public SKColor getBackgroundColor() {
		synchronized (this) {
			return backgroundColor;
		}
	}

	public void setBackgroundColor(SKColor color) {
		synchronized (this) {
			backgroundColor.assignByValue(color);
		}
	}

	public void setAnchorPoint(SKPoint anchor) {
		synchronized (this) {
			anchorPoint.assignByValue(anchor);
		}
	}

	public SKPoint getAnchorPoint() {
		synchronized (this) {
			return anchorPoint;
		}
	}

	public SKView getView() {
		synchronized (this) {
			return view;
		}
	}

	protected void setView(SKView view) {
		synchronized (this) {
			this.view = view;
		}
	}

	public SKPoint convertGLPointrom(SKPoint point) {
		synchronized (this) {
			return null;
		}
	}

	public SKPoint convertPointTo(SKPoint point) {
		synchronized (this) {
			return null;
		}
	}

	protected long getCurrentTime() {
		synchronized (this) {
			return view.getCurrentTime();
		}
	}

	public SKSize getSize() {
		synchronized (this) {
			return size;
		}
	}

	public void setSize(SKSize size) {
		synchronized (this) {
			this.size.assignByValue(size);
		}
	}

	@Override
	public void onDrawFrame(GLRenderer renderer, int width, int height) {
		synchronized (this) {
			drawChildren(renderer, width, height);
		}
	}

	public void movedToView(SKView view) {
		synchronized (this) {

		}
	}

	public void movedFromView(SKView view) {
		synchronized (this) {

		}
	}

	void changedSize(SKSize oldSize) {
		synchronized (this) {

		}
	}

	@Override
	public boolean onTouch(View v, MotionEvent event) {
		synchronized (this) {
			return false;
		}
	}
}

File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKScene.kt
package br.com.insanitech.spritekit

import br.com.insanitech.spritekit.graphics.SKColor
import br.com.insanitech.spritekit.graphics.SKPoint
import br.com.insanitech.spritekit.graphics.SKSize

open class SKScene : SKEffectNode {
    var anchorPoint: SKPoint = SKPoint()
        set(value) { field.point.assignByValue(value.point) }

    var size: SKSize = SKSize()
        set(value) {
            val oldSize = SKSize(field)
            field.size.assignByValue(value.size)
            this.didChangeSize(oldSize)
        }

    var scaleMode: SKSceneScaleMode = SKSceneScaleMode.AspectFill

    var backgroundColor: SKColor = SKColor(0.15f, 0.15f, 0.15f, 1.0f)
        set(value) { field.color.assignByValue(value.color) }

    var view: SKView? = null
        internal set

    constructor(size: SKSize) {
        this.size = size
        this.sceneDidLoad()
    }

    open fun didChangeSize(oldSize: SKSize) {

    }

    open fun sceneDidLoad() {

    }

    open fun willMove(fromView: SKView) {

    }

    open fun didMove(toView: SKView) {

    }

    open fun update(currentTime: Long) {

    }

    open fun didEvaluateActions() {

    }

    open fun didFinishUpdate() {

    }

    override fun addChild(node: SKNode) {
        super.addChild(node, false)
        this.movedToScene(this)
    }

    override fun insertChild(node: SKNode, index: Int) {
        super.insertChild(node, index, false)
        this.movedToScene(this)
    }

    override fun convertPointForCandidateParentIfNeeded(p: SKPoint, parent: SKNode): SKPoint {
        return if (parent == this) p else parent.convertFrom(p, this)
    }
}
